presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PropertyDerivations.java

```diff
@@ -438,11 +440,59 @@ public ActualProperties visitJoin(JoinNode node, List<ActualProperties> inputPro
                             .unordered(true)
                             .build();
                 case FULL:
-                    // We can't say anything about the partitioning scheme because any partition of
-                    // a hash-partitioned join can produce nulls in case of a lack of matches
-                    return ActualProperties.builder()
-                            .global(probeProperties.isSingleNode() ? singleStreamPartition() : arbitraryPartition())
-                            .build();
+                    if (SystemSessionProperties.isDeriveFullJoinProperties(session)) {
+                        // the join keys might not be output by the join node, so we need to see if we can express the partitioning scheme
+                        // for the join in terms of its outputs
+                        probeProperties = probeProperties.translate(column -> filterIfMissing(node.getOutputSymbols(), column));
+                        buildProperties = buildProperties.translate(column -> filterIfMissing(node.getOutputSymbols(), column));
+
+                        if (!probeProperties.getNodePartitioning().isPresent() || !buildProperties.getNodePartitioning().isPresent()) {
+                            return ActualProperties.builder()
```

Comment:
maybe extract these properties before `if (SystemSessionProperties.isDeriveFullJoinProperties(session)) {`
as some kind of `defaultProperties` or `fallbackProperties` (I'm not sure about the name)